---
title: "S.cerevisiae 2017: Collecting annotation data (20180606)."
author: "atb abelew@gmail.com"
date: "`r Sys.Date()`"
output:
 html_document:
  code_download: true
  code_folding: show
  fig_caption: true
  fig_height: 7
  fig_width: 7
  highlight: default
  keep_md: false
  mode: selfcontained
  number_sections: true
  self_contained: true
  theme: readable
  toc: true
  toc_float:
    collapsed: false
    smooth_scroll: false
---

<style>
  body .main-container {
    max-width: 1600px;
  }
</style>

```{r options, include=FALSE}
if (!isTRUE(get0("skip_load"))) {
  library("hpgltools")
  tt <- devtools::load_all("~/hpgltools")
  knitr::opts_knit$set(progress=TRUE,
                       verbose=TRUE,
                       width=90,
                       echo=TRUE)
  knitr::opts_chunk$set(error=TRUE,
                        fig.width=8,
                        fig.height=8,
                        dpi=96)
  old_options <- options(digits=4,
                         stringsAsFactors=FALSE,
                         knitr.duplicate.label="allow")
  ggplot2::theme_set(ggplot2::theme_bw(base_size=10))
  ver <- "20180606"
  previous_file <- "index.Rmd"

  tmp <- try(sm(loadme(filename=paste0(gsub(pattern="\\.Rmd", replace="", x=previous_file), "-v", ver, ".rda.xz"))))
  rmd_file <- paste0("01_annotation-v", ver, ".Rmd")
}
```

# Annotation version: `r ver`

## Genome annotation input

There are a few methods of importing annotation data into R.  I will attempt
some of them in preparation for loading them into the S.cerevisiae RNASeq data.

### AnnotationHub: loading OrgDb

AnnotationHub is a newer service and has promise to be an excellent top-level
resource for gathering annotation data.  Its organization is peculiar, one
connects to the database with AnnotationHub() and downloads lists of sqlite
databases via query().  The primary problem with AnnotationHub is that it seems
difficult to keep up with the changes in their database -- so one must
double-check the downloaded data to be sure that it is still actually data and
not just a message saying 'Public'.

Once it does download data, one may use normal orgdb functions to query it.

```{r data_input_genome}
tmp <- sm(library(AnnotationHub))
ah <- sm(AnnotationHub())
orgdbs <- sm(query(ah, "OrgDb"))
sc_orgdb <- sm(query(ah, c("OrgDB", "Saccharomyces"))) ##   AH49589 | org.Sc.sgd.db.sqlite
sc_orgdb <- sc_orgdb[[1]]
sc_orgdb

## Holy crap it worked!
sc_annotv1 <- load_orgdb_annotations(
  sc_orgdb,
  fields=c("alias", "description", "entrezid", "genename", "sgd"))
sc_annotv1 <- sc_annotv1[["genes"]]
```

### TxDb

In yeast, the transcript database is not super-useful, given that there are only
80 some genes wit introns, but it does provide a quick way to get transcript lengths.

```{r scerevisiae_txdb}
tt <- please_install("TxDb.Scerevisiae.UCSC.sacCer3.sgdGene")
tmp <- library(TxDb.Scerevisiae.UCSC.sacCer3.sgdGene)
sc_txdb <- TxDb.Scerevisiae.UCSC.sacCer3.sgdGene
```

# Loading a genome

There is a non-zero chance we will want to use the actual genome sequence along with these
annotations.  The BSGenome packages provide that functionality.

```{r scerevisiae_bsgenome}
tt <- sm(please_install("BSgenome.Scerevisiae.UCSC.sacCer3"))
```

# Loading from biomart

A completely separate and competing annotation source is biomart.  Biomart
provides programmatic access to the tremendous quantity of ensembl data.
Their data is organized in some weird ways, including tables with thousands of
available columns; but man is it awesome to have so much available stuff to
sift through and learn about.

```{r scerevisiae_biomart}
sc_annotv2 <- sm(load_biomart_annotations("scerevisiae"))
sc_annotv2 <- sc_annotv2[["annotation"]]
head(sc_annotv2)
sc_ontology <- sm(load_biomart_go("scerevisiae"))
sc_ontology <- sc_ontology[["go"]]
head(sc_ontology)
```

# Read a gff file

In contrast, it is possible to load most annotations of interest directly from the gff files used in
the alignments.  The main problem with gff data is that the format is incredibly
inconsistent; but it is often the most direct way to go from the IDs of the
genome to some immediately useful data.

```{r genome_input}
## The old way of getting genome/annotation data
sc_gff <- "reference/scerevisiae.gff.gz"
sc_gff_annotations <- load_gff_annotations(sc_gff, type="gene")
rownames(sc_gff_annotations) <- make.names(sc_gff_annotations$transcript_name, unique=TRUE)
head(sc_gff_annotations)
```

# Putting the pieces together

In the following block we create an expressionset using the sample sheet and the
annotations.

Annoyingly, the gff annotations are keyed in a peculiar fashion.  Therefore I
need to do a little work to merge them.

In the following block, I spend a little time setting up locations by
chromosome/start/end and using those to merge the gff data and biomart data,
thus solving the problem of inconsistent IDs.  It is worth noting that I split
this process between the genes on the + strand and those on the - strand because
the definitions of 'beginning of gene' and 'start' mean different things:
'beginning of gene' refers to the location with respect to the start codon and
is used by biomart, 'start' refers to the location with respect to the beginning
of the chromosome and is used by the gff data.

```{r create_expt}
## Start by making locations for the biomart data
sc_annotv2[["fwd_location"]] <- paste0(sc_annotv2[["chromosome"]], "_", sc_annotv2[["start"]])
sc_annotv2[["rev_location"]] <- paste0(sc_annotv2[["chromosome"]], "_", sc_annotv2[["end"]])
## Do the same for the gff annotations
sc_gff_annotations[["fwd_location"]] <- paste0(sc_gff_annotations[["seqnames"]], "_",
                                               sc_gff_annotations[["start"]])
sc_gff_annotations[["rev_location"]] <- paste0(sc_gff_annotations[["seqnames"]], "_",
                                               sc_gff_annotations[["end"]])
sc_gff_annotations[["gff_rowname"]] <- rownames(sc_gff_annotations)
## Now merge them.
sc_fwd_annotations <- merge(sc_annotv2, sc_gff_annotations, by="fwd_location")
sc_rev_annotations <- merge(sc_annotv2, sc_gff_annotations, by="rev_location")
colnames(sc_fwd_annotations) <- c("location","transcriptID","geneID", "Description",
                                  "Type", "length", "chromosome", "strand.x", "start.x",
                                  "end.x", "location.x", "seqnames",
                                  "start.y", "end.y", "width", "strand.y", "source", "type",
                                  "score", "phase", "exon_number", "gene_id", "ID", "p_id",
                                  "protein_id", "transcript_id", "transcript_name", "tss_id",
                                  "seqedit", "location.y", "gff_rowname")
colnames(sc_rev_annotations) <- colnames(sc_fwd_annotations)
sc_all_annotations <- rbind(sc_fwd_annotations, sc_rev_annotations)
rownames(sc_all_annotations) <- make.names(sc_all_annotations[["gff_rowname"]], unique=TRUE)
sc_all_annotations <- sc_all_annotations[, c("transcriptID", "geneID", "Description", "Type",
                                             "length", "chromosome", "strand.x", "start.x", "end.x",
                                             "tss_id")]
colnames(sc_all_annotations) <- c("transcriptID", "geneID", "Description", "Type", "length",
                                  "chromosome", "strand", "start", "end", "tss_id")
sc_all_annotations[["location"]] <- paste0(sc_all_annotations[["chromosome"]], "_",
                                           sc_all_annotations[["start"]], "_",
                                           sc_all_annotations[["end"]])
```

### Make the expressionset

The function 'create_expt() gathers the annotation data, metadata, and counts,
and invokes the various R functions to create an expressionset.  This function
is more complex than it should be, primarily because I go to some effort to
accept a large array of inputs including:  raw data frames of counts vs. sets of
filenames of counts/hdf5/tpm/etc, data frames of metadata vs. excel sheets with
varying columns, and arbitrary annotation data which may include all, some, or
none of the included genes.

```{r expt}
sc2_expt <- create_expt(
  metadata="sample_sheets/all_samples.xlsx",
  gene_info=sc_all_annotations,
  file_column="bt2file")
knitr::kable(head(exprs(sc2_expt$expressionset)))
knitr::kable(head(fData(sc2_expt$expressionset)))
knitr::kable(head(pData(sc2_expt$expressionset)))
```

```{r saveme}
if (!isTRUE(get0("skip_load"))) {
  pander::pander(sessionInfo())
  message(paste0("This is hpgltools commit: ", get_git_commit()))
  this_save <- paste0(gsub(pattern="\\.Rmd", replace="", x=rmd_file), "-v", ver, ".rda.xz")
  tmp <- sm(saveme(filename=this_save))
}
```
